runner/runner.py

```diff
@@ -32,16 +33,27 @@ def add_position_to_block_names(pipeline: Pipeline) -> Pipeline:
     entire_pipeline = pipeline.children()
 
     def add_position(block: Union[List[Element], Element], position: int, prefix: str):
-        prefix += f"{position}-"
         if isinstance(block, List):
-
+            if position > 0:
+                prefix += f"{position - 1}-"
             for i, child in enumerate(block):
-                add_position(child, position + i, prefix)
-        else:
+                add_position(child, i, prefix)
+        elif not isinstance(block, DataSource):
             block.id += f"{prefix}{position}"
 
-    for child in entire_pipeline:
-        add_position(child, 0, "-")
+    add_position(entire_pipeline, 0, "-")
+
+    return pipeline
+
+
+def replace_duplicates_with_copy(pipeline: Pipeline) -> Pipeline:
+    entire_pipeline = flatten(pipeline.children())
+    seen = []
+    for block in entire_pipeline:
+        if block not in seen:
+            seen.append(block)
+        else:
+            block = deepcopy(block)
```

Comment:
If I am correct, this is needed because of the following example:
- we create ``modelA``
- we put ``modelA`` in multiple places
- we train the pipeline, and everytime ``modelA`` gets mutated

This can introduce unwanted behaviour.
Maybe we are already copying it somewhere that I'm not aware of, let me know.

Reply:
I'm testing it and the way I do it seems to be currently wrong.

Reply:
good catch!
we already deepcopy the model inside SKLearn, so we don't share the parameters, but that probably won't be possible with HF.
this affects transformations as well.

I think we should deepcopy all blocks at the initalization of the Pipeline object. It's a simple array comprehension, and does it at the root, not later (like here).

Reply:
I've opened an alternative PR for this.

the problem here is that you're simply re-assigning the block in the local scope, but that doesn't change it in the "children" array. and it'd be quite messy to do that!